#459134 [C#テキスト]7章演習問題

kurosu-being opened the pull request:
第7章 配ディクショナリの操作

ディクショナリの操作（初期化、要素の格納方法、ディクショナリへの変換）について学習しました。

kurosu-being committed:
#459134 [C#テキスト]7章演習問題

Practicce7-1-1を実装しました。

kurosu-being committed:
#459134 [C#テキスト]7章演習問題

Practice7-1-2を実装しました。
Practice7-2-1のファイルを作成しました。

kurosu-being committed:
#459134 [C#テキスト]7章演習問題

Practice7-2-1の途中まで追加しました。

kurosu-being committed:
#459134 [C#テキスト]7章演習問題

Practice7-2-1を仮に実装しました。

kurosu-being committed:
#459134 [C#テキスト]7章演習問題

Practice7-2-2を実装しました。
Practice7-2-3を実装しました。

kurosu-being committed:
#459134 [C#テキスト]7章演習問題

Practice7-2-3のProgram.csの名前を変更しました。

kurosu-being committed:
#459134 [C#テキスト]7章演習問題

問題文に対応した実装内容になるようにコード全体の削除、追加を行いました。
Practice7-2-4を実装しました。
Practice7-2-3、Practice7-2-4において、パスをユーザーに打ち込んでもらう形式に変更しました。

kurosu-being committed:
#459134 [C#テキスト]7章演習問題

Practice7-2のAbbreviationsクラスにクラスのコメントを追加しました。
Practice7-2-4において、規定のフォーマット通りに整形しました。

kurosu-being committed:
#458736 [C#テキスト]6章演習問題

Practice7-2全体のインデクサ、メソッドの引数名を変更しました。
Practice7-2全体の不要なコメントを削除しました。

kurosu-being committed:
#458736 [C#テキスト]6章演習問題

空白の削除などフォーマットの修正を行いました。
Practice7-2のすべてにおいて、Abbreviationクラスのプロパティにthisを追加しました。

kurosu-being committed:
#459134 [C#テキスト]7章演習問題

Practice7-2-1、Practice7-2-2において、パスを引数で受け取る形式に変更しました。

kurosu-being committed:
.editorconfigファイルを追加しました。

r-nishiguchi-being commented on 章末問題7/Practice7-2-1/Abbreviations.cs:
```diff
@@ -0,0 +1,59 @@
+        /// </summary>
+        /// <param name="vKey">省略語</param>
+        /// <returns>キーに対応する日本語</returns>
+        public string this[string vKey] => FDict.ContainsKey(vKey) ? FDict[vKey] : null;
```
Dictionaryからキーを使って値を取り出すとき、Gaiaの開発ではTryGetValue()が推奨されます。
記法は以下のように書きます。
```
FDict.TryGetValue(vKey, out int wValue);
return wValue;
```
利点は以下です。
①戻り値がboolのため、vKeyがあればどうする。なければどうするということを決めることがでます
②vKeyがなかった場合、wValueにはデフォルトの値が入ってきます。
③FDict[vKey]より安全にValueを取り出すことができます。

kurosu-being replied:
ご指摘ありがとうございます！
TryGetValue()を使い、修正いたしました。
TryGetValue()で確実に値を取得するように書くことが商品のエラーを防ぐことに繋がるのですね。
勉強になります🙇‍♀️

r-nishiguchi-being replied:
OKです

kurosu-being committed:
#459134 [C#テキスト]7章演習問題

Practice7-2-1以降のAbbreviationsクラスのカウントコンストラクタを、TryGetValueメソッドで書き直しました。

k-miyahara-being commented on 章末問題7/Practice7-1-1/Program7-1-1.cs:
```diff
@@ -0,0 +1,31 @@
+
+            wInputString = wInputString.ToUpper();
+            var wCharCount = new Dictionary<char, int>();
+            foreach (char wCharacter in wInputString) {
```
元の文字列が上書きされてしまっています。
元の文字列は維持するようにしてください。

kurosu-being replied:
ご指摘ありがとうございます。
foreachの中でwInputString.ToUpper()を直接渡す形に変更し、元の文を書き換えないよう修正いたしました。

k-miyahara-being commented on 章末問題7/Practice7-1-1/Program7-1-1.cs:
```diff
@@ -0,0 +1,31 @@
+                if (char.IsLetter(wCharacter)) {
+                    wCharCount[wCharacter] = wCharCount.ContainsKey(wCharacter) ? wCharCount[wCharacter] + 1 : 1;
+                }
+            }
```
この7行の処理は1行で実現することができます。
ToDictionary()を使った方法で書き換えてみてください。
（7-1-2の解答はそのままでよいです。）

kurosu-being replied:
ご指摘ありがとうございます！
ToDictionary()を使い、一行で処理を書き直しました。

k-miyahara-being commented on 章末問題7/Practice7-1-2/Program7-1-2.cs:
```diff
@@ -0,0 +1,22 @@
+            foreach (char wCharacter in wInputString) {
+                if (char.IsLetter(wCharacter)) {
+                    wCharCount[wCharacter] = wCharCount.ContainsKey(wCharacter) ? wCharCount[wCharacter] + 1 : wCharCount[wCharacter] = 1;
+                }
```
ガード節を使ってみてください。

また、14行目はもう少しコンパクトに書けます。
常に1ずつ増える場合はインクリメントを使ったほうが、数字を書き換えられる等によるバグが埋め込みされづらくなります。
`wCharCount[wCharacter] = 1`も右辺の`1`だけで十分ですね。

kurosu-being replied:
ご指摘ありがとうございます！
ガード節を使い修正いたしました。
インクリメントの前置後置がどちらかが怪しくかったので式で書いてしまいましたが、数字を書き換えられる等によるバグを埋め込まない為にも正しく理解したいと思います。

k-miyahara-being commented on 章末問題7/Practice7-2-1/Abbreviations.cs:
```diff
@@ -0,0 +1,64 @@
+        public string ToAbbreviation(string vJapanese) => FDict.FirstOrDefault(x => x.Value == vJapanese).Key;
+
+        /// <summary>
+        /// 日本語の位置を引数に与え、それが含まれる要素（Key、Value）をすべて取り出すメソッド
+        /// </summary>
+        /// <param name="vSubstring">日本語の位置</param>
```
日本語の"一部" ではないですかね？
記憶が定かじゃないですが、テキストが誤字っていたような…

kurosu-being replied:
ご指摘ありがとうございます！
位置→一部に変更いたしました。
ニュアンスで位置ともいえるか・・・と己を納得させてしまいました😅

k-miyahara-being commented on 章末問題7/Practice7-2-3/Program7-2-3.cs:
```diff
@@ -0,0 +1,35 @@
+                //削除後の要素数を取得
+                wCount = wAbbreviations.Count;
+                Console.WriteLine($"削除後の要素の数: {wCount}");
+            } else {
+                Console.WriteLine("存在しないパスです");
+            }
```
ファイルが存在するかどうかのチェックとても良いです！
ここもガード節を使ってみましょう。

kurosu-being replied:
ありがとうございます！💪
ガード節を使いファイルが存在しない場合早期に処理を返すように修正しました。

k-miyahara-being commented on 章末問題7/Practice7-2-4/Abbreviations.cs:
```diff
@@ -0,0 +1,80 @@
+        /// <returns>指定した省略後を削除したディクショナリ</returns>
+        public bool Remove(string vKey) => FDict.Remove(vKey);
+
+        /// <summary>
+        /// ３文字の省略語だけ取り出しコンソールに表示するメソッド
+        /// </summary>
+        public void DisplayThree() {
+            foreach (var wAbbr in FDict.Where(x => x.Key.Length == 3)) {
+                Console.WriteLine($"{wAbbr.Key}={wAbbr.Value}");
+            }
+        }
```
メソッドの役割が「該当の要素を取り出す」と「出力する」の2つになってしまっています。
例えば、取り出した要素を使って別の操作をしたい、取り出した要素の値を書き換えたいといったとき、
毎回余計な出力処理が行われてしまいます。

メソッドは役割がなるべくシンプルで、それ単体でどこからでも使いまわせるようになっていることが理想的です。

kurosu-being replied:
ご指摘ありがとうございます！
メソッドの機能が単機能になるようにDisplayThreeメソッド修正しました。

> メソッドは役割がなるべくシンプルで、それ単体でどこからでも使いまわせるようになっていることが理想的です。

メソッドを単機能にすることにおいて、どこまでが単機能かが分かりづらいと感じていたのですが、それ単体でどこからでも使いまわせるという観点がありませんでした。勉強になります！

`

k-miyahara-being replied:
メソッドの機能にあわせて、命名も変えられているのでGoodです！

k-miyahara-being reviewed (changes requested):
ほぼOKです！
少し確認をお願いします。

kurosu-being committed:
#459134 [C#テキスト]7章演習問題

Practice7-1-1 において、19行目から25行目までの処理をLINQを用いて一行で書き直しました。
Practice7-1-2 において、ガード節で文字でない場合は処理を飛ばすように処理を纏めました。
Practice7-2-1～Practice7-2-4ににおいて、誤字を修正いたしました。（位置→一部）

kurosu-being committed:
459134 [C#テキスト]7章演習問題

Practice7-1-2において、インクリメントで書き直しました。
Practice7-2-3において、ガード節で書き直しました。
Practice7-2-4において、メソッドが単機能になるように修正しました。

kurosu-being committed:
Practice7-2-4において、不要なメソッドを削除しました。

k-miyahara-being commented on 章末問題7/Practice7-2-3/Program7-2-3.cs:
```diff
@@ -0,0 +1,36 @@
+                Console.WriteLine("存在しないパスです");
+                return;
+            }
+                var wAbbreviations = new Abbreviations(wFilePath);
+
+                //いくつかの要素を追加
+                wAbbreviations.Add("KM", "黒子 真紀");
+                wAbbreviations.Add("HAS", "八太 亜美 さん");
+                wAbbreviations.Add("MKS", "宮原 啓樹 さん");
+
+                //現在の要素数を取得
+                int wCount = wAbbreviations.Count;
+                Console.WriteLine($"要素の数: {wCount}");
+
+                //要素を削除
+                bool wRemoved = wAbbreviations.Remove("KM");
+                Console.WriteLine($"KM を削除: {wRemoved}");
+
+                //削除後の要素数を取得
+                wCount = wAbbreviations.Count;
+                Console.WriteLine($"削除後の要素の数: {wCount}");
```
細かいですが、インデントがずれています。

kurosu-being replied:
失礼しました。コード整形をして修正いたしました。
コード整形の確認をさらに強化します。

kurosu-being committed:
#459134 [C#テキスト]7章演習問題

Practice7-1-1,7-2-2,7-2-3,7-2-4において、コード整形を行いました。

k-miyahara-being reviewed (approved):
OKです！

yuki-tanaka-being commented on 章末問題7/Practice7-1-1/Program7-1-1.cs:
```diff
@@ -0,0 +1,26 @@
+
+            var wCharCount = wInputString.ToUpper().Where(char.IsLetter).GroupBy(x => x).ToDictionary(x => x.Key, x => x.Count());
+
+            foreach (var wKeyValuePair in wCharCount.OrderBy(x => x.Key)) {
+                Console.WriteLine($"'{wKeyValuePair.Key}':{wKeyValuePair.Value}");
```
本題とは逸れますが、[`KeyValuePair<TKey, TValue>`](https://learn.microsoft.com/ja-jp/dotnet/api/system.collections.generic.keyvaluepair-2?view=netframework-4.7.2) は [分解](https://learn.microsoft.com/ja-jp/dotnet/csharp/fundamentals/functional/deconstruct) という書き方ができます。
```csharp
foreach ((char wChar, int wCount) in wCharCount.OrderBy(x => x.Key)) {
    Console.WriteLine($"'{wChar}':{wCount}");
}
```
まだ学習してないかもですが、タプルと呼ばれるものを扱うときに知っていると便利です。
利点は、簡単に分かりやすく変数名を付けられることです。
[未確認飛行 C - 複合型の分解](https://ufcpp.net/study/csharp/datatype/deconstruction/#deconstruction) のさわり部分を読むと便利さがわかるかと思います。
タプルに限らず KeyValuePair も Key とか Value って結局何が入っているの？というのが変数名を付けることで分かりやすくなります。

kurosu-being replied:
ありがとうございます！
分解という書き方は初めて知りましたが、タプルを使うとKeyValuePair のKeyとValueがコード上で明確になり、他の場所でもアドバイス頂いた「コードの表現力」という点が大きく改善されますね！

自分でも使ってみたいと思いコードを書きなおそうとしたのですが、ご提示いただいたコードを使用すると

> 'KeyValuePair<char, int>' に 'Deconstruct' の定義が含まれておらず、型 'KeyValuePair<char, int>' の最初の引数を受け付けるアクセス可能な拡張メソッド 'Deconstruct' が見つかりませんでした。

というエラーメッセージが出ました。
そこで2点質問があります。

①このエラーを解消する為にはどのように修正すればよいのでしょうか？
②このエラーの原因を「使用している.NET Frameworkではまだ 'KeyValuePair<char, int>' に 'Deconstruct' の定義が実装されていない為使うことが出来ない」と考えたのですが、これは正しいのでしょうか？
（こちらに関してはこの[サイト](https://qiita.com/RyotaMurohoshi/items/dda4b2d623ddc9e0f6ff#:~:text=Deconstruct%E3%83%A1%E3%82%BD%E3%83%83%E3%83%89%E3%81%8CKeyValuePair%3CTKey%2C%20Tvalue%3E%E3%81%AA%E3%81%A9%E3%81%AB%E8%BF%BD%E5%8A%A0%E3%81%95%E3%82%8C%E3%81%A6%E3%80%81ValueTuple%E3%81%BF%E3%81%9F%E3%81%84%E3%81%AB%E5%88%86%E8%A7%A3%E5%AE%A3%E8%A8%80%E3%81%A7%E3%81%8D%E3%82%8B%E3%82%88%E3%81%86%E3%81%AB%E3%81%AA%E3%82%8B%20C%23.NET%20Last%20updated%20at%202019-03-02%20Posted,%E3%81%95%E3%81%A6%E3%80%81%E3%81%93%E3%81%AE%E5%88%86%E8%A7%A3%E3%81%A7%E3%81%99%E3%81%8C%20ValueTuple%20%E5%9E%8B%E3%81%A7%E3%81%AA%E3%81%8F%E3%81%A6%E3%82%82%20Deconstruct%20%E3%81%A8%E3%81%84%E3%81%86%E5%90%8D%E5%89%8D%E3%81%AE%E3%82%A4%E3%83%B3%E3%82%B9%E3%82%BF%E3%83%B3%E3%82%B9%E3%83%A1%E3%82%BD%E3%83%83%E3%83%89%E3%81%8B%E6%8B%A1%E5%BC%B5%E3%83%A1%E3%82%BD%E3%83%83%E3%83%89%E3%82%92%E6%8C%81%E3%81%A3%E3%81%A6%E3%81%84%E3%82%8C%E3%81%B0%E4%BD%BF%E3%81%86%E3%81%93%E3%81%A8%E3%81%8C%E3%81%A7%E3%81%8D%E3%81%BE%E3%81%99%E3%80%82%20%E6%A8%99%E6%BA%96%E3%83%A9%E3%82%A4%E3%83%96%E3%83%A9%E3%83%AA%E3%81%AE%E5%9E%8B%E3%81%AB%20Deconstruct%20%E3%81%A8%E3%81%84%E3%81%86%E5%90%8D%E5%89%8D%E3%81%AE%E3%82%A4%E3%83%B3%E3%82%B9%E3%82%BF%E3%83%B3%E3%82%B9%E3%83%A1%E3%82%BD%E3%83%83%E3%83%89%E3%80%81%E3%82%82%E3%81%97%E3%81%8F%E3%81%AF%E6%8B%A1%E5%BC%B5%E3%83%A1%E3%82%BD%E3%83%83%E3%83%89%E3%81%8C%E5%8A%A0%E3%82%8F%E3%82%8A%E5%88%86%E8%A7%A3%E3%81%8C%E3%81%A7%E3%81%8D%E3%82%8B%E3%82%88%E3%81%86%E3%81%AB%E3%81%AA%E3%82%8A%E3%81%BE%E3%81%97%E3%81%9F%E3%80%82)を見て結論付けました）

お忙しいところ恐縮ですが、 ご教授いただけると幸いです🙇‍♀️

yuki-tanaka-being replied:
ごめんなさい。調べてもらった通りでした。
KeyValuePair の分解は、.NET Core 2.0 以上からじゃないと標準サポ―トしていませんでした。
タプルの分解は、C# 7 以降から対応しているので使えます。

KeyValuePair の分解を .NET Core 2.0 未満で使う方法ですが、以下のクラスとメソッドを用意して、使いたいクラスで参照できるようにしておけば使えるようになります。
```csharp
    public static class KeyValuePairExtensions {

        /// <summary>
        /// 分解構文を有効にします。
        /// </summary>
        /// <example>
        /// <code><![CDATA[
        /// foreach ((int wCD, string wName) in wDictionary) {
        /// }
        /// ]]></code>
        /// </example>
        public static void Deconstruct<TKey, TValue>(this KeyValuePair<TKey, TValue> vPair, out TKey vKey, out TValue vValue) {
            vKey = vPair.Key;
            vValue = vPair.Value;
        }
    }
```
標準で用意されていないだけで、自前で実装すれば使えるようになります。
Gaia Cloudのソースには、この実装が含まれていたので使えていただけでした😂

kurosu-being replied:
自身のVisualStudioでも正常に動作するのを確認できました！調べて頂きありがとうございます。
分解の書き方からGaia Cloudのソースについても知ることが出来て嬉しいです。扱えるようにクラスを用意してあったのですね😳
サポートのバージョンで出来ることが増えたり、便利な書き方が出来るように独自で拡張メソッドやクラスを追加していたりと、色んな方々のコードの可読性を上げていこうとする試みを垣間見た気がします。
ありがとうございます！

yuki-tanaka-being commented on 章末問題7/Practice7-2-1/Abbreviations.cs:
```diff
@@ -0,0 +1,64 @@
+    /// 省略語のクラス
+    /// </summary>
+    class Abbreviations {
+        private Dictionary<string, string> FDict = new Dictionary<string, string>();
```
初期化以外でインスタンスを変更しないクラス変数は `readonly` を付けるとバグを埋め込みにくい堅牢なプログラムになります。

kurosu-being replied:
ご指摘ありがとうございます！
readonlyを付け修正いたしました。

yuki-tanaka-being replied:
修正ありがとうございます！
動作的に問題はありませんが、記載する順番は `private readonly` が一般的かなと思います。
更に static を付ける場合は `private static readonly` の順番になります。
製品コードを書く時は、周りの実装見ながら真似ればよいです！

kurosu-being replied:
ご指摘ありがとうございます！
記載する順番を修正いたしました。
フォーマットを守ることもコードの可読性には重要ですね。
製品コードや他者のコードに手を加えるときなど、周りの実装を真似るように心がけたいと思います。

yuki-tanaka-being commented on 章末問題7/Practice7-2-1/Abbreviations.cs:
```diff
@@ -0,0 +1,64 @@
+    /// 省略語のクラス
+    /// </summary>
+    class Abbreviations {
+        private Dictionary<string, string> FDict = new Dictionary<string, string>();
```
Dictionary の変数名は KeyToValue がおすすめです。
Key と Value 部分は、実際に何が入るかを入れます。
例えば、ユーザーIDがキーでユーザー名が値だった場合、`IDToName` みたいな感じです。
変数名から Key と Value が何か分かるようになり、コード全体の可読性が上がるのと、バグを埋め込みにくくなります。

kurosu-being replied:
ご指摘ありがとうございます！
FDictをFAbbraviationToJapaneseに変更いたしました。
（長い名前ですが、スコープの広いフィールド変数なので分かりやすさの方を重視すべきと判断しました）

yuki-tanaka-being replied:
コードは問題ないですが、ここの投稿だけ `w` 始まりになってますね😂

kurosu-being replied:
失礼しました😱コメントを修正いたしました。

yuki-tanaka-being commented on 章末問題7/Practice7-2-1/Abbreviations.cs:
```diff
@@ -0,0 +1,64 @@
+        /// <returns>キーに対応する日本語</returns>
+        public string this[string vKey] {
+            get {
+                FDict.TryGetValue(vKey, out string wValue);
+                return wValue;
```
レビューで TryGetValue を使う方法を教えてもらっているかと思いますが、こちらも一行で書くことができます。
`public string this[string vKey] => FDict.TryGetValue(vKey, out string wValue) ? wValue : null;`
良くある書き方なので、覚えておいて損はないです！

ちなみに、TryGetValue の戻り値の bool を使わないコードがあったら、キーが無い時の考慮が漏れているか、すでにキーがあることが保証されており TryGetValue を使う必要性がないか、のどちらかを疑うとよいです。

kurosu-being replied:
ご指摘ありがとうございます！
ご提示いただいたコードで修正いたしました。
TryGetValueや他の戻り値がbool型のメソッドについて、考慮が漏れていないか特に注意したいと思います。

yuki-tanaka-being commented on 章末問題7/Practice7-2-1/Abbreviations.cs:
```diff
@@ -0,0 +1,64 @@
+        /// </summary>
+        /// <param name="vSubstring">日本語の一部</param>
+        /// <returns>含まれる要素（Key、Value）</returns>
+        public IEnumerable<KeyValuePair<string, string>> FindAll(string vSubstring) {
+            foreach (var wItem in FDict) {
+                if (wItem.Value.Contains(vSubstring))
+                    yield return wItem;
+            }
+        }
```
メソッドを利用する側のことを考えると、戻り値の型は `KeyValuePair<string, string>` ではなく、`(string Abbreviation, string Japanese)` のタプルが使いやすいです。
型をみるだけで、戻り値が何か分かるので可読性がよいです。
また、分解せずにタプルのまま変数に取った場合も、`x.Abbreviation` と `x.Japanese` となるので分かりやすいです。
KeyValuePair だと、`x.Key` と `x.Value` となり、抽象度が高くて何かわからないです。

---

可読性の話を幾つか書いていますが、プログラムにおけるコードの表現力に着目すると面白いです。
コメントではなく、如何にコードで他者へ意図を伝えられるかを意識してみてください。
コードだけで意図を表現できる場所ってそもそもどこだろうか？
メソッド名、引数の型、引数名、戻り値の型、変数名、アルゴリズム、など沢山あります。
一つ一つは些細なものもありますが、すべてに意味を込めて書かれたコードと、まったく意識していないコードとでは、読みやすさが段違いになります。
まずは、沢山の表現の仕方を吸収するところから始めるのがよいです。
まずは、アウトプットよりインプット重視！

kurosu-being replied:
ご指摘ありがとうございます！
タプルの扱いについて分かり次第こちらも修正したいと思います。

> メソッド名、引数の型、引数名、戻り値の型、変数名、アルゴリズム、など沢山あります。
> 一つ一つは些細なものもありますが、すべてに意味を込めて書かれたコードと、まったく意識していないコードとでは、読みやすさが段違いになります。

同じことではないかもしれませんが、同じ機能だからとコピペで切り貼りしたコードと、一から必要だと思いながら書いたコードでは読みやすさが全然違うと感じます。一行一行がコンピューターへの命令と思って意味を込めるのが大切ですね。


> まずは、沢山の表現の仕方を吸収するところから始めるのがよいです。
> まずは、アウトプットよりインプット重視！

書籍や先輩方のレビューから吸収すべく頑張ります！💪

yuki-tanaka-being commented on 章末問題7/Practice7-2-3/Abbreviations.cs:
```diff
@@ -0,0 +1,70 @@
+        /// <summary>
+        /// コンストラクタ
+        /// </summary>
+        public Abbreviations(string vFilePath) {
+            var wLines = File.ReadAllLines(vFilePath);
+            FDict = wLines.Select(x => x.Split('=')).ToDictionary(x => x[0], x => x[1]);
+        }
```
`Abbreviations` クラスから [I/O](https://e-words.jp/w/I-O.html) 処理を分割すると、[単一責任](https://ja.wikipedia.org/wiki/%E5%8D%98%E4%B8%80%E8%B2%AC%E4%BB%BB%E3%81%AE%E5%8E%9F%E5%89%87) のクラスになります。
元のテキストを読んでいないので今回のお題に反しているかもしれませんが、ちゃんとした製品コードを書くなら気にした方がよい話になります。

ここでの I/O 処理は、コンストラクタでファイルパスを受け取って、ファイルから文字列を読み取っている処理です。
現在の `Abbreviations` を使うには、ファイルからテキストを読み取ることが強要されます。
もし、単純に省略語と日本語の辞書というクラスが使いたいだけの人がいた場合、必ずファイルから生成することが強いられます。つらいです。
もし、.NET 標準の Dictionary クラスを new するのに、必ずファイルからデータを取ってくる必要があった場合を想像してみてください。つらいです。
というお話になります。

じゃあ空のコンストラクタも用意すれば解決するかというと、それも違います。
ファイルからではなくネットワーク経由で初期データを用意したい要求がでてきた場合を想像してみてください。
コンストラクタを増やすか？
それでも対応はできますが、初期化方法が増える度にクラスが肥大化していきます。
しかも本来の、辞書としての役割には全く関係ない部分で変わります。つらいです。
単一責任のクラスにすることで、複雑度が下がり、メンテナンスしやすいコードになります。

解決策は、シンプルに  I/O 処理を外に出して、`Abbreviations` クラスは、引数なしのコンストラクタと、初期化時にデータをセットできるように `IDictionary<TKey,TValue>` を引数に取るコンストラクタを実装します。
オブジェクト内のデータは、使う側が自由に設定できるようにします。

kurosu-being replied:
ご指摘ありがとうございます！
Practice7-2-3において、Abbreviations クラスに空のコンストラクタとIDictionary<TKey,TValue> を引数に取るコンストラクタに変更いたしました。
（正しく実装できているかまだ不安の為、一時的にPractice7-2-3のみで実装しています）
汎用的なコードになるようにDictionary<TKey, TValue> ではなくIDictionary<TKey,TValue>を引数にとるのは以前もご指摘頂いた部分ですね😀
コンストラクタに何を設定すべきか、製品として利用する時どういうことがおこりそうかなど、先輩方に解説頂きながら沢山勉強になりました。ありがとうございます！

yuki-tanaka-being replied:
正しく実装できてますよー👍 
これで Abbreviations クラスは再利用しやすいものとなりました。

今度は逆に、ファイルを読み込んで Abbreviations クラスを生成する部分を main に書くことになりました。
main が I/O 処理を持つことになりましたが、今回はこれで構いません。
もし、この処理を使いまわしたくなった場合は、ファイルから Abbreviationsクラスを生成する役割をもつクラスを作成します。
そうすることで、生成とモデルを分離しつつ、どちらも再利用（組み合わせ）しやすいものが出来上がります。
プログラムが大きくなるにつれて、このようにリファクタリングをしていくことが大切になってきます。

yuki-tanaka-being reviewed (commented):
レビューしました。
ご確認お願いします。

kurosu-being commented:
GitHubのテストです（演習問題とは無関係です）
[リンク](https://qiita.com/RyotaMurohoshi/items/dda4b2d623ddc9e0f6ff#:~:text=Deconstruct%E3%83%A1%E3%82%BD%E3%83%83%E3%83%89%E3%81%8CKeyValuePair%3CTKey%2C%20Tvalue%3E%E3%81%AA%E3%81%A9%E3%81%AB%E8%BF%BD%E5%8A%A0%E3%81%95%E3%82%8C%E3%81%A6%E3%80%81ValueTuple%E3%81%BF%E3%81%9F%E3%81%84%E3%81%AB%E5%88%86%E8%A7%A3%E5%AE%A3%E8%A8%80%E3%81%A7%E3%81%8D%E3%82%8B%E3%82%88%E3%81%86%E3%81%AB%E3%81%AA%E3%82%8B%20C%23.NET%20Last%20updated%20at%202019-03-02%20Posted,%E3%81%95%E3%81%A6%E3%80%81%E3%81%93%E3%81%AE%E5%88%86%E8%A7%A3%E3%81%A7%E3%81%99%E3%81%8C%20ValueTuple%20%E5%9E%8B%E3%81%A7%E3%81%AA%E3%81%8F%E3%81%A6%E3%82%82%20Deconstruct%20%E3%81%A8%E3%81%84%E3%81%86%E5%90%8D%E5%89%8D%E3%81%AE%E3%82%A4%E3%83%B3%E3%82%B9%E3%82%BF%E3%83%B3%E3%82%B9%E3%83%A1%E3%82%BD%E3%83%83%E3%83%89%E3%81%8B%E6%8B%A1%E5%BC%B5%E3%83%A1%E3%82%BD%E3%83%83%E3%83%89%E3%82%92%E6%8C%81%E3%81%A3%E3%81%A6%E3%81%84%E3%82%8C%E3%81%B0%E4%BD%BF%E3%81%86%E3%81%93%E3%81%A8%E3%81%8C%E3%81%A7%E3%81%8D%E3%81%BE%E3%81%99%E3%80%82%20%E6%A8%99%E6%BA%96%E3%83%A9%E3%82%A4%E3%83%96%E3%83%A9%E3%83%AA%E3%81%AE%E5%9E%8B%E3%81%AB%20Deconstruct%20%E3%81%A8%E3%81%84%E3%81%86%E5%90%8D%E5%89%8D%E3%81%AE%E3%82%A4%E3%83%B3%E3%82%B9%E3%82%BF%E3%83%B3%E3%82%B9%E3%83%A1%E3%82%BD%E3%83%83%E3%83%89%E3%80%81%E3%82%82%E3%81%97%E3%81%8F%E3%81%AF%E6%8B%A1%E5%BC%B5%E3%83%A1%E3%82%BD%E3%83%83%E3%83%89%E3%81%8C%E5%8A%A0%E3%82%8F%E3%82%8A%E5%88%86%E8%A7%A3%E3%81%8C%E3%81%A7%E3%81%8D%E3%82%8B%E3%82%88%E3%81%86%E3%81%AB%E3%81%AA%E3%82%8A%E3%81%BE%E3%81%97%E3%81%9F%E3%80%82)

kurosu-being committed:
#459134 [C#テキスト]7章演習問題

Practice7-1-1において、KeyValuePair<TKey, Tvalue>を分解の書き方で一時的に書き直しました。※エラーがあります

kurosu-being committed:
#459134 [C#テキスト]7章演習問題

Practice7-1-1において、不要なusingを削除しました。
Practice7-2-3において、コンストラクタを二つ用意しメインメソッドで読み込みを行うことで、Abbreviationsクラスから読み取り機能を除外しました。

kurosu-being committed:
#459134 [C#テキスト]7章演習問題

Practice7-2-2において、FDict→FAbbreviationToJapaneseに変更しました。
Practice7-2-3において、FDict→FAbbreviationToJapaneseに変更しました。
Practice7-2-3において、FDict→FAbbreviationToJapaneseに変更しました。
Practice7-2-4において、FDict→FAbbreviationToJapaneseに変更しました。

kurosu-being committed:
#459134 [C#テキスト]7章演習問題

Practice7-1-1において、レビュー前の状態に一時的に戻しました。
Practice7-2-1~7-2-4において、Abbreviationsクラスのインデクサを式本体で書くように変更し、条件分岐を追加しました。

kurosu-being committed:
#459134 [C#テキスト]7章演習問題

Practice7-2-3において、データをセットするコンストラクタを修正しました。

kurosu-being committed:
#459134 [C#テキスト]7章演習問題

Practice7-2以降において、AbbreviationsクラスのFAbbreviationToJapaneseのアクセス修飾子の順番を変更しました。

kurosu-being committed:
#459134 [C#テキスト]7章演習問題

Practice7-1-1において、分解構文が使えるようにKeyValueExtensions.csを追加しました。

yuki-tanaka-being commented on 章末問題7/Practice7-2-4/Abbreviations.cs:
```diff
@@ -0,0 +1,73 @@
+        /// <summary>
+        /// ３文字の省略語だけ取り出すメソッド
+        /// </summary>
+        public IEnumerable<string> GetThreeAbbr() {
+            var wThreeAbbr = FAbbreviationToJapanese.Where(x => x.Key.Length == 3).Select(x => $"{x.Key}={x.Value}");
+            return wThreeAbbr;
+        }
```
メソッド名の付け方について、確定したコレクションを返すか、遅延評価される列挙を返すか、で異なる表現をするのがおすすめです。
列挙を返すのに `GetXXX` は使わない方がよいです。代わりに `EnumerateXXX` などを使います。
Get は一般的に定まっているものを返すイメージです。
また、`IEnumerable<T>` を返すならすべて `EnumerateXXX` でもないので注意です。
（列挙を受け取って特定の条件でフィルタリングした列挙を返すメソッドなら Filter など）

■参考
* [Directory.EnumerateFiles メソッド](https://learn.microsoft.com/ja-jp/dotnet/api/system.io.directory.enumeratefiles?view=netframework-4.7.2)
* [Directory.GetFiles メソッド](https://learn.microsoft.com/ja-jp/dotnet/api/system.io.directory.getfiles?view=netframework-4.7.2)

どちらも指定したフォルダ内のファイル情報を返してくれメソッドですが、戻り値の型でメソッド名が異なるのが分かるかと思います。
理由もあります。上記のDirectoryクラスを例にします。
すべてのファイル情報が欲しいならどちらを使っても大差ないです。
もし、特定のファイル名を含むものがあるか知りたいとした場合、EnumerateFiles の方は見つかった時点で、残りのファイルは一切参照しません。
GetFiles の方は、最初にすべてのファイル情報へアクセスします。
もし、フォルダ内に数万ファイルあったとして、どちらがパフォーマンスが良いか考えると、EnumerateFiles の方が期待値が高いです。
そういった動作の違いが、メソッド名だけで使う側に分かるようになっているのが利点です。
ちゃんと遅延評価されているというのが分かるのもポイントです。
※フォルダ内のファイルを参照する処理内で、ファイルを追加するような処理が入る場合、遅延評価は正しく動作しません。（.NET だと実行時エラーが発生するはずです）
　先に、全ファイル情報を取得してからループを回す必要があります。

kurosu-being replied:
ご指摘ありがとうございます！
GetThreeAbbr→EnumerateThreeAbbrに修正いたしました。
取得する戻り値が列挙型の時EnumerateXXXという名付けがセオリーであることは知りませんでした。
確かに、配列と列挙型は似ているようでコード上での振る舞い方が大きく異なる為、どちらなのか明示する方が便利ですね。
以後使っていきたいと思います！

yuki-tanaka-being reviewed (commented):
なぜか 7.2.4 を見ていなかったので追加でレビューしました。
ご確認お願いします。

kurosu-being committed:
#459134 [C#テキスト]7章演習問題

Practice7-2-4において、GetThreeAbbr→EnumerateThreeAbbrに修正いたしました。

yuki-tanaka-being commented on 章末問題7/Practice7-2-4/Abbreviations.cs:
```diff
@@ -0,0 +1,73 @@
+        /// <summary>
+        /// ３文字の省略語だけ取り出すメソッド
+        /// </summary>
+        public IEnumerable<string> EnumerateThreeAbbr() {
+            var wThreeAbbr = FAbbreviationToJapanese.Where(x => x.Key.Length == 3).Select(x => $"{x.Key}={x.Value}");
+            return wThreeAbbr;
+        }
```
今回のお題に対しては今の実装で問題ありません。
クラス設計の観点では戻り値の型を `IEnumerable<(string, string)>` にしたり、省略語と日本語を変数に持つクラスを作成して `IEnumerable<newClass>` にする方のがおすすめです。
理由は、使う側からみて汎用性が高い方が使いやすく再利用性が高いからです。
例えば、このメソッドが2か所以上から呼ばれるとして、ある場所では `省略語=日本語` と表示したいし、別のある場所では `[省略語]: 日本語` と表示したいとなったときに、今の実装では対応できません。
メソッド側でフォーマットを構築してしまっているからです。
構造化されたデータに比べて文字列データは扱いづらくなってしまいます。
戻り値の型を決める際にぜひ思い出してみてください。
